enhancements/cvo/object-removal-manifest-annotation.md

```diff
@@ -0,0 +1,134 @@
+---
+title: use-manifest-annotation-for-object-removal
+authors:
+  - "@jottofar"
+reviewers:
+  - "@LalatenduMohanty"
+  - "@wking"
+  - "@crawford"
+  - "@sdodson"
+approvers:
+  - "@crawford"
+  - "@sdodson"
+creation-date: 2020-08-03
+last-updated: 2020-08-04
+status: implementable
+---
+
+# Use Manifest Annotation For Object Removal
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Currently CVO managed object removal is handled by jobs.
+Jobs have many immutable components, so they aren't a great match for CVO managed manifests.
+This enhancement replaces that approach with a new manifest annotation requesting the CVO delete the in-cluster object instead of creating/reconciling it.
+This will provide a more straightforward way to remove content.
+
+## Motivation
+
+There should be a straightforward way to remove CVO managed objects.
+
+### Goals
+
+* Users will be able to remove any of the currently managed CVO objects by simply modifying the object's existing manifest and adding the new delete annotation.
+
+### Non-Goals
+
+* This enhancement does not add any new objects for managment by CVO, therefore the new delete annotation only applies to the currently supported manifests and any that may be added in the future.
+
+## Proposal
+
+When the following annotation appears in a CVO supported manifest and is set to "true" the associated object will be removed from the cluster by the CVO.
+Values other than `true` will result in a CVO error and no further processing of the manifest.
+
+```
+apiVersion: apps/v1
+...
+metadata:
+...
+  annotations:
+    release.openshift.io/delete: "true"
+```
+
```

Comment:
please describe the impact of these resources on status.  Are you waiting for finalization or only marking with a deletiontimestamp.

Reply:
Naively i would say finalization is required in order to ensure that old versions of the process are gone deterministically.

Reply:
> Naively i would say finalization is required in order to ensure that old versions of the process are gone deterministically.

This sort of blocking is what we started to move away from with #400.  I'd rather just see `DeletionTimestamp` and say "ah, API server is already on this" and continue through the manifest graph.  And when it's not set, assume that a successful `Delete` call is sufficient to queue the removal and continue through the manifest graph.  Do you have an example where we'd need to block on finalization to avoid the CVO tripping over itself?

Reply:
If we don't confirm deletion by the CVO, then who's going to confirm the deletion actually took place?  Objects can persist indefinitely with a deletion timestamp if finalizers aren't removed by appropriate entities, etc.  If we have the CVO block and fail if the object doesn't go away in a timely manner (maybe keep track of 'things I deleted' and do a wait at the end of the run to minimize runtime).  Another consideration for making it synchronous is if any component needs that object to be gone to function correctly.

If the delete is async, something could behave differently in production on a user's cluster (eg, they have modified some child object in a way that results in finalizers being stuck) and then determining what the problem is would be that much harder.  EG, operator X is failing, it's not clear why, would be clear if we failed on the manifest deletion step.

Reply:
> If we don't confirm deletion by the CVO, then who's going to confirm the deletion actually took place?

I would still have the CVO set `Upgradeable=False` or some such to block further minor-version bumps if deletion-annotated resources still existed (i.e. had not been finalized).  Do you see a situation where a resource flagged for deletion in a given z stream could break other components in that same z stream?

Reply:
Added finalization section.